httpbp/client_middlewares.go

```diff
@@ -0,0 +1,194 @@
+
+type roundTripper func(req *http.Request) (*http.Response, error)
+
+func (f roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config Config, middlewares ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+
+	defaults := []ClientMiddleware{
+		CircuitBreaker(config),
+	}
+
+	middlewares = append(middlewares, defaults...)
+
+	return &http.Client{
+		Transport: applyMiddlewares(transport, middlewares...),
+	}
+}
+
+// CircuitBreaker is a middleware that prevents sending requests that are likely
+// to fail based on a configurable failure ratio based on total failures and
+// requests.
+func CircuitBreaker(config Config) ClientMiddleware {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return &gobreaker.Settings{
+				Interval: 60 * time.Second,
+				ReadyToTrip: func(counts gobreaker.Counts) bool {
+					failureRatio := float64(counts.TotalFailures) / float64(counts.Requests)
+					return counts.Requests >= config.CircuitBreaker.MinRequests && failureRatio >= config.CircuitBreaker.MinFailureRatio
+				},
+			}
+		},
+	}
+	breakers := &sync.Map{}
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (*http.Response, error) {
+			host := req.URL.Hostname()
+
+			// new circuit breakers should rarely get allocated: the pool reduces
+			// garbage collector overhead
+			newBreaker := pool.Get()
+			breaker, loaded := breakers.LoadOrStore(host, newBreaker)
+			if loaded {
+				pool.Put(newBreaker)
+			}
+
+			resp, err := breaker.(*gobreaker.CircuitBreaker).Execute(func() (interface{}, error) {
+				resp, err := next.RoundTrip(req)
+				if err != nil {
+					return nil, err
+				}
+				// circuit break on any HTTP 5xx code
+				if resp.StatusCode >= http.StatusInternalServerError {
+					err = resp.Body.Close()
+					if err != nil {
+						return nil, err
+					}
+					return nil, err
+				}
+				return resp, nil
+			})
+			if err != nil {
+				return nil, err
+			}
+			return resp.(*http.Response), nil
+		})
+	}
+}
+
+func applyMiddlewares(transport http.RoundTripper, middlewares ...ClientMiddleware) http.RoundTripper {
+	// add middlewares in reverse so the first in the list is the outermost
+	for i := len(middlewares) - 1; i >= 0; i-- {
+		transport = middlewares[i](transport)
+	}
+	return transport
+}
+
+// Retries provides a retry middleware.
+func Retries(config retry.Config) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			retrybp.Do(req.Context(), func() error {
+				resp, err = next.RoundTrip(req)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, retry.Attempts(1))
```

Comment:
we should only add `retry.Attemps(1)` if there's no other options specified in the config, like what we did in `thriftbp`: https://github.com/reddit/baseplate.go/blob/88dcf6c19330435fa8c6766f1b97c9c40adfc75a/thriftbp/client_middlewares.go#L110-L112